Explicit voyageai embed support

laszukdawid opened the pull request:
## Description

Adds explicit support for VoyageAI embedding. These embedding functions were already in the code, under `anthropic.py`, but there wasn't explicit support for using them via `EMBEDDING_MODEL`.

## Related Issues

None that I've seen.

## Changes Made

- Externalize `voyageai` logic into `voyageai.py` from `anthropic.py`
- Add validation for `voyageai` as `args.embedding_model` value

## Checklist

- [x] Changes tested locally
- [x] Code reviewed
- [ ] Documentation updated (if necessary)
- [ ] Unit tests added (if applicable)

## Additional Notes

N/A

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds first-class VoyageAI embeddings with a new provider module and integrates it into server binding/dispatch; removes VoyageAI code from Anthropic.
> 
> - **Embeddings**:
>   - **New provider**: Introduces `lightrag/llm/voyageai.py` with `voyageai_embed` (retry, AsyncClient) wrapped via `wrap_embedding_func_with_attrs` (`embedding_dim=1024`, `max_token_size=16000`) and helper `get_available_embedding_models`.
>   - **Server integration**: Adds `"voyageai"` to supported `embedding_binding` values and dispatch in `create_optimized_embedding_function` to call `lightrag.llm.voyageai.voyageai_embed`.
> - **Refactor/Cleanup**:
>   - Removes VoyageAI embedding code and `numpy` import from `lightrag/llm/anthropic.py`.
> - **DX**:
>   - Improves unsupported embedding error message to include the invalid value.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 9afc59956f71c279a453b2df5719ebf1228970b4. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

laszukdawid committed:
chore/support-voyageai-embed-directly: feat: voyageai embed support

laszukdawid committed:
chore/support-voyageai-embed-directly: fix: revert uv.lock changes

laszukdawid committed:
Merge branch 'main' into chore/support-voyageai-embed-directly

laszukdawid committed:
chore/support-voyageai-embed-directly: chore: resolve comments

danielaskdd commented:
@codex review

oscampo commented:
Hi
I’ve just performed a test on LightRAG v1.4.13 to verify the current state of Voyage AI integration.
Despite correctly configuring the environment variables (including VOYAGEAI_API_KEY and the proper endpoint host), the system consistently returns a 400 Bad Request. The logs confirm that the internal Rerank wrapper is still hardcoding the top_n parameter in the JSON payload:

`ERROR: Rerank API error 400: {"detail":"...Argument 'top_n' is not supported by our API"}`

This proves that the current 'Cohere-proxy' approach is fundamentally incompatible with Voyage AI's schema. I have a 200M token quota available and I am ready to run extensive integration tests to help you validate a definitive fix that properly maps top_n to top_k (or removes it) for Voyage. This is definitely a necessary fix for the community. Let's get this merged!

danielaskdd commented:
Thank you for your interest and contributions to LightRAG. Given the significant changes introduced in version 1.5, we will review this PR following the official release of v1.5.

In the meantime, if you wish to integrate embedding and re-ranking models from alternative providers, we recommend deploying a LiteLLM proxy server to handle API routing. LiteLLM can normalize APIs from various vendors into OpenAI and Cohere-compatible formats, thereby enabling LightRAG to seamlessly leverage a wide range of third-party models.

LarFii committed:
fix(voyage): pass EMBEDDING_MODEL through to voyageai_embed dispatch

The voyageai branch in `create_optimized_embedding_function` only forwarded
`texts`, `api_key` and `embedding_dim` to `voyageai_embed`, so any value of
`EMBEDDING_MODEL` configured by the user (e.g. `voyage-3-lite`,
`voyage-code-3`, etc.) was silently ignored and the function defaulted to
`voyage-3`. This caused dim mismatches against vector stores configured for
`voyage-3-lite`'s 512 dimensions.

Mirror the openai/jina/gemini branches and conditionally include `model` in
the kwargs.

Made-with: Cursor

LarFii committed:
fix(voyage): accept legacy VOYAGE_API_KEY env var + broaden retries

Several improvements to lightrag/llm/voyageai.py:

1. Backward compatibility: read both `VOYAGE_API_KEY` (the name VoyageAI's
   own SDK and docs use) and `VOYAGEAI_API_KEY`. Existing deployments that
   relied on the previous `anthropic_embed` function used `VOYAGE_API_KEY`,
   so they would have broken with `ValueError` on first run otherwise.

2. Retry coverage: extend the tenacity decorator to cover `Timeout`,
   `ServiceUnavailableError`, `ServerError` and `TryAgain` from
   `voyageai.error`. The previous coverage only handled `RateLimitError`
   and `APIConnectionError`, so transient 502/503 / SDK timeouts surfaced
   immediately to the caller.

3. Default `truncation=True` (matches the SDK default explicitly rather
   than relying on `None == default`).

4. Move `import voyageai` to module top-level (was redundantly re-imported
   inside the function body).

5. Fix `truncation` doc + remove misleading "Ignore api_key" debug log
   (the function actually prefers the explicit `api_key` argument).

Made-with: Cursor

LarFii committed:
fix(anthropic): keep anthropic_embed importable as deprecated shim

Removing `anthropic_embed` from lightrag/llm/anthropic.py is a clean refactor
inside this repo (no internal callers reference it after this PR), but it
silently breaks any downstream code that has

    from lightrag.llm.anthropic import anthropic_embed

Add a thin compatibility wrapper that:

  * stays importable under the old name,
  * delegates to `lightrag.llm.voyageai.voyageai_embed.func`,
  * emits `DeprecationWarning` with the migration target.

Slated for removal in a future major release.

Made-with: Cursor

LarFii committed:
test(voyage): add unit tests for voyageai_embed and anthropic_embed shim

Six fully-mocked tests covering:

  * `model` kwarg is forwarded to the SDK (would have caught the regression
    where the server's voyageai dispatch dropped the model parameter);
  * both `VOYAGE_API_KEY` and `VOYAGEAI_API_KEY` env vars are accepted;
  * a clear `ValueError` is raised when neither is configured;
  * `input_type` kwarg reaches the SDK (so callers can drive query/document
    selection);
  * `lightrag.llm.anthropic.anthropic_embed` remains importable and emits
    `DeprecationWarning`.

Pure unit tests; no live API calls. Runs in <1s under pytest-asyncio.

Made-with: Cursor

LarFii commented:
@laszukdawid thanks for cleaning up the misplaced VoyageAI code in `anthropic.py` — the direction is right and `AsyncClient` is a better fit than the previous sync client. I reviewed the PR end-to-end with a real Voyage key (covered direct calls, env-var handling, and the full `ainsert` + `aquery` path through the binding) and pushed a small set of fixes directly to your branch (since you opted in to maintainer edits). Happy to drop them or rework anything if you'd rather take a different approach.

## What I pushed (4 commits on top of `2752b01f`)

| Commit | Type | Why |
|---|---|---|
| `14354ec` `fix(voyage): pass EMBEDDING_MODEL through to voyageai_embed dispatch` | regression fix | The `voyageai` branch in `create_optimized_embedding_function` only forwarded `texts`, `api_key`, `embedding_dim` — `model` was dropped, so `EMBEDDING_MODEL=voyage-3-lite` (or anything other than `voyage-3`) silently had no effect and produced 1024-dim vectors regardless of what the vector store was sized for. Mirrors the openai/jina/gemini branches. (Earlier commit `45700cdf` did include `model=model`; it appears to have been lost in a later squash.) |
| `2d5db52` `fix(voyage): accept legacy VOYAGE_API_KEY env var + broaden retries` | backward-compat + reliability | (a) Read both `VOYAGE_API_KEY` (the name VoyageAI's own SDK + docs use) and `VOYAGEAI_API_KEY`. The previous `anthropic_embed` on `main` used `VOYAGE_API_KEY`, so users upgrading would have hit `ValueError` immediately. Empirically reproduced before the fix. (b) Extended `tenacity` retry to also cover `Timeout`, `ServerError`, `ServiceUnavailableError`, `TryAgain` from `voyageai.error`; previously transient 502/503 surfaced to the caller and aborted the whole batch. (c) Misc: explicit `truncation=True` default to match SDK, hoist `import voyageai` to module top, drop the misleading `"Ignore api_key"` debug log. |
| `58099e0` `fix(anthropic): keep anthropic_embed importable as deprecated shim` | breaking-change mitigation | Removing `anthropic_embed` cleanly is fine inside this repo (no internal callers), but breaks `from lightrag.llm.anthropic import anthropic_embed` for any downstream user. Added a thin shim that delegates to `voyageai_embed.func` and emits `DeprecationWarning`. Slated for removal in a future major release. |
| `9ddd5fc` `test(voyage): add unit tests for voyageai_embed and anthropic_embed shim` | testing | 6 fully-mocked tests (no live API): model forwarding, both env-var names, missing-key error, `input_type` forwarding, deprecation shim. Would have caught the model-dropping regression at PR time. Runs in ~1s. |

Net stat: `+162 / -31`.

## Verification

- 6/6 new unit tests pass; pre-existing `tests/test_chunking.py` (43 tests) still pass.
- Real Voyage e2e against the patched code:
  - direct call with `model=voyage-3-lite` returns `(N, 512)` shape.
  - `VOYAGE_API_KEY` (legacy name) only set → embedding works.
  - `from lightrag.llm.anthropic import anthropic_embed` works and emits `DeprecationWarning`.

## Safety net

I created a `backup/pr-2484-prereview-2026-04-25` ref on `HKUDS/LightRAG` pointing at your previous head `2752b01f`, so the original commit can be restored at any time with:

```bash
git push origin <commit-sha>:refs/heads/chore/support-voyageai-embed-directly --force
# or with the backup ref:
gh api -X PATCH repos/laszukdawid/LightRAG/git/refs/heads/chore/support-voyageai-embed-directly \
  -f sha=2752b01f1242eb24e12cc5bd59e83f18434f537b -F force=true
```

## Items I deliberately did NOT touch

- **Hardcoded `embedding_dim=1024` in the decorator**: I tried removing it but `EmbeddingFunc.embedding_dim` is required and removing the default would be a wider refactor (touches `lightrag/utils.py`). I added a docstring note instead so users on `voyage-3-lite` know they need to set `EMBEDDING_DIM=512`.
- **Coupling with #2560 (asymmetric/task-aware embeddings)**: When both PRs land, the auto-merge succeeds but the voyageai dispatch doesn't receive the new `context`/prefix plumbing. Voyage's `input_type="query"|"document"` is the perfect fit for #2560's design — recommend a follow-up PR after both merge that adds `supports_asymmetric=True` to `voyageai_embed` and maps `context` → `input_type` in the server dispatch. I can open that follow-up, or you can — happy either way.

If anything here doesn't fit your intent, just push back / revert any of the commits and I'll adapt. Thanks again for the contribution!

LarFii committed:
style: apply ruff-format and remove unused imports

CI uses ruff v0.6.4 (pinned in .pre-commit-config.yaml). Reformat to match
that version's defaults and drop unused ``os`` / ``AsyncMock`` imports.

Made-with: Cursor

laszukdawid commented:
Thanks @LarFii ! Honestly, this isn't my call; I'm just a bystander who needed Voyage to work and thought I'd share here. If you have merge permissions - I'd love if you merged this. My code base is quite different from LightRAG at this point but merging this would be a signal that maybe contributions are accepted, and I'd be happy to contribute.
